glean-core/src/upload/mod.rs

```diff
@@ -0,0 +1,357 @@
+        }
+    }
+}
+
+/// Creates requests for ping upload and process responses from them.
+pub struct PingUploadHelper {
+    /// A queue of ping UUIDs to upload
+    queue: PingsQueue,
+    /// The path to the pending_pings directory
+    pending_pings_dir: PathBuf,
+}
+
+impl PingUploadHelper {
+    /// Create a new PingUploadHelper.
+    ///
+    /// Starts an empty queue to be filled with the pending pings
+    /// file paths that should be processed and uploaded next.
+    pub fn new(data_path: &Path) -> Self {
+        Self {
+            queue: PingsQueue::new(),
+            pending_pings_dir: data_path.join("pending_pings"),
+        }
+    }
+
+    /// Creates a `PingRequest` for the first possible ping in the queue.
+    ///
+    /// If one of the files is malformed or generates errors, it will
+    /// keep trying while there are still pending pings.
+    pub fn create_next_ping_request(&self) -> Option<PingRequest> {
+        let mut ping = self.get_next_ping();
+        // As long as there are pings in the queue this will
+        // keep trying to create a `PingRequest` from it.
+        while ping.is_some() {
+            // Safe unwrap, we state that ping "is_some" before entering the loop
+            let (uuid, path) = ping.unwrap();
+            let request = self.process_ping_file(uuid, path);
+            if request.is_some() {
+                return request;
+            }
+            ping = self.get_next_ping();
+        }
+        None
+    }
+
+    /// Process the next ping, if there are any in the queue.
+    ///
+    /// Will try **once** to re-process the pending_pings folder
+    /// in case the queue is empty.
+    fn get_next_ping(&self) -> Option<(Uuid, PathBuf)> {
+        let mut attempts = 0;
+        loop {
+            if let Some(uuid) = self.queue.process() {
+                // Create the ping file path from the UUID
+                let path = self.pending_pings_dir.join(uuid.to_string());
+                return Some((uuid, path));
+            }
+
+            if let Err(e) = self.process_pings_dir() {
+                log::error!("Error processing pings directory! {}", e);
+            }
+
+            // Will only attempt once to process the
+            // pending pings directory before giving up.
+            if attempts > 0 {
+                break;
+            } else {
+                attempts += 1;
+            }
+        }
+        None
+    }
+
+    /// Process and add all valid ping files to queue.
+    ///
+    /// Any files that don't match the UUID regex will be deleted
+    /// to prevent files from polluting the pings directory.
+    ///
+    /// Any directories will also be deleted. There should only be
+    /// files inside the `pending_pings` directory.
+    fn process_pings_dir(&self) -> Result<()> {
+        for entry in self.pending_pings_dir.read_dir()? {
+            if let Ok(entry) = entry {
+                let path = entry.path();
+
+                if path.is_dir() {
```

Comment:
Is this a new behaviour?

Reply:
Hm, just checked. Yes it is, for some reason I wrote it down as something this function should do. If you think it doesn't make sense, I can remove it.

Reply:
Let's keep it simple for now and just reproduce what we currently have with respect to this. We can always add later.

Reply:
I just remembered where I got this from, https://github.com/mozilla/glean/blob/master/glean-core/tests/common/mod.rs#L115-L118